keps/sig-scalability/0001-enhance-network-performance-benchmark-testcases.md

```diff
@@ -0,0 +1,152 @@
+With wide array of network functions being containerized, the need to analyse and 
+monitor the networking performance metrics such as latency, throughput etc has arisen.
+
+We have come across 2 test frameworks which have limited testcases for network performance benchmarking.
+
+a) E2E (test\e2e\network\networking_perf.go): This framework has only 2 testcases for measuring throughput.
+
+b) netperf (https://github.com/kubernetes/perf-tests/tree/master/network/benchmarks/netperf) : This framework 
+   measures only the throughput between client-server pods for various MSS(Maximum Segment size). 
+   Other performance metrics such as latency , loss-rate , response time are not measured.
+
+### Goals
+
+The primary goal of this proposal is to enhance the testcases to cover various 
+networking scenarios :
+
+a)  IPv4 , IPv6 , Dual-Stack.
+b)  Protocols: TCP, UDP, HTTP/HTTPS.
+c)  With services such as ClusterIP , NodePort etc.
+
+We intend to submit tests to measure the following network performance metrics for various 
+scenarios such as different services (ClusterIP/NodePort), with IPv4/IPv6 configured etc.
+
+a)  Throughput (For TCP traffic)
+b)  Loss rate (For UDP traffic)
+c)  Latency/Jitter (For UDP/HTTP/HTTPS)
+d)  Response time (For HTTP/HTTPs)
+
+Our important goal is to make sure these benchmarking testcases are run periodically,
+so that any degradation in performance is detected in early phase of feature delivery.
+
+### Non-goals
+
+Benchmarking and measurements with intermediate services like ingress controllers, load-balancers 
+are out of scope of this proposal.
+
+### Challenges and Open Questions [optional]
+
+1) Which is the suitable hosting framework for proposed tests ? 
+2) Periodic execution of these proposed tests ?
+
+
+## Proposal
+
+### Test Scenarios
+
+As captured in the link https://docs.google.com/document/d/1_760ptAHBY_c56slWhjl6PJQOrf__vQFJXkL58ewg1M/edit
+we propose to have 2 broad categories based on how client PODs can communicate with server PODs.
+
+The testcase will cover different protocols such as TCP, UDP, HTTP/HTTPS with IPv4/IPv6 addresses configured.
+
+The testcases will cover different client-to-server pod number ratio (i.e 1:N , 1:1 , N:1)
+
+### Tools
+
+We intend to use the following tools for testing in addition to iPerf/netperf already being used
+
+a)  T-Rex: T-Rex is opensource traffic generator tool made available by CISCO.
+    It will be used for measuring the UDP maximum throughput (Kpps), RTT latency (min,max,avg).
+
+b)  Siege: Siege is an open source benchmark utility. 
+    It will be used for measuring HTTP/HTTPs metrics such as total number of hits recorded, 
+    bytes transferred, response time, concurrency, and return status. 
+
+	
+### Execution Environment
+
+We propose to enhance the testcases in either of the 2 test frameworks i.e E2E or netperf.
+
+Our recommendation is to include these testcases in kubernetes CI framework,so that these
```

Comment:
what requirements do you need? number of nodes, VMs resources, ... 
I know that Trex works fine for AWS, how about GCE, are there any NIC/network limitations?
Any specific CNI plugin to test with?

Reply:
Regarding requirements ,  
a) Worker nodes : The worker nodes to run the client-server pods. The number might vary based on the testcases and number of pods we want to run on them.
b) vCPUs : I believe , here the worker nodes are run on the VMs. So these VMs should have enough vCPUs assigned , so that the tools we run inside the POD can make use these vCPUs. We can pin the vCPUs to the tools , so that they run efficienlty.

Regarding the Trex, i am not sure of its limitations on various provider cloud platforms like AWS , GCE. 
I have used Trex only in my local k8s cluster.

Regarding CNIs, we plan to have generic test cases written which are agnostic to CNI.
CNI configuration should be preparatory step.
But initially i would like to test with Calico , Multus CNI.

Reply:
@mm4tt are any "networking specifics" in the cluster used by the tests? I mean, special network drivers and these kind of things. If you have run any dpdk test before is a good thing, because this will assure Trex compatibiliy.

Reply:
@mm4tt  are calico and multus CNIs available for us to run these testcases ?

Reply:
I think I'm actually against extending k8s e2e tests - I would prefer to use either of the two other existing ones. We're generally going into direction to not use them for performance/scalability related testing.

Re networking specific - both are about tests and are independent of how the cluster is configured. You should be able to configure your clusters as you wish independently from the test you want to run on it.

Reply:
@wojtek-t , do we have any CNI specific testcaes in sig-scalability or its not in the scope of sig-scalability? 
As per my knowledge of scalability tests, they are run on GCP cluster. Which CNI is the default one on GCP ?